Unbreak Perl build on legacy Darwin systems

sevan opened the pull request:
On modern macOS the compiler defaults to '-Werror,-Wimplicit-function-declaration' This breaks the build on code which calls a function without a prototype (invalid in C99).
Turning the check off with '-Wno-error=implicit-function-declaration' breaks the build of Perl itself on legacy versions of Darwin with compilers which do not recognise the flags.
Removing it allows Perl to once again build on Mac OS X as old as 10.4 with GCC 4.0.1.

Resolves #18902

Leont commented:
Changes under `cpan/` should be filed [upstream](https://github.com/Perl-Toolchain-Gang/ExtUtils-MakeMaker/) instead.

Otherwise I approve of this change.

sevan commented:
Moved PR into draft mode until other PR for MakeMaker is merged.

haarg commented:
The changes to EUMM should be removed from this PR, since that will be handled separately.

Making this change, along with the EUMM change, is likely to break a number of XS modules on macOS. I agree that we should change this, but I'm not sure about the timing.

sevan committed:
Unbreak Perl build on legacy Darwin systems
On modern macOS the compiler defaults to '-Werror,-Wimplicit-function-declaration'
This breaks the build on code which calls a function without a prototype
(invalid in C99).
Turning the check off with '-Wno-error=implicit-function-declaration'
breaks the build of Perl itself on legacy versions of Darwin with
compilers which do not recognise the flags.
Removing it allows Perl to once again build on Mac OS X as old as 10.4
with GCC 4.0.1.
Bump version and document.
This requires a similar change to cpan/ExtUtils-MakeMaker/lib/ExtUtils/MM_Darwin.pm
which has already been handled, separately.
https://github.com/Perl-Toolchain-Gang/ExtUtils-MakeMaker/pull/444

sevan commented:
I've removed `cpan/ExtUtils-MakeMaker/lib/ExtUtils/MM_Darwin.pm`

sevan commented:
[Smoke test result with unpatched blead](https://perl5.test-smoke.org/report/5033206)

sevan commented:
[Smoke test result with patched blead](https://perl5.test-smoke.org/file/log_file/5033391)

Ignore the failures, I didn't bump the rev so t/porting/cmp_version.t failed, cpan/ExtUtils-MakeMaker/t/04-xs-rpath-darwin.t  fails but the test shouldn't be running on Darwin 8.
There is an issue with cpan/Test-Simple/t/Test2/behavior/ipc_wait_timeout.t hanging which is why the smoke test took so long.  I had to run 'kill -9' several times over the past few days.

jkeenan commented:
> [Smoke test result with patched blead](https://perl5.test-smoke.org/file/log_file/5033391)
> 
> Ignore the failures, I didn't bump the rev so t/porting/cmp_version.t failed, cpan/ExtUtils-MakeMaker/t/04-xs-rpath-darwin.t fails but the test shouldn't be running on Darwin 8. There is an issue with cpan/Test-Simple/t/Test2/behavior/ipc_wait_timeout.t hanging which is why the smoke test took so long. I had to run 'kill -9' several times over the past few days.

There has been no discussion in this ticket in more than a year, and in that time the code has acquired several merge conflicts.  Those will have to be resolved before we can proceed.

sevan commented:
I'll take a look at updating over the weekend.
